core/modules/user.py

```diff
@@ -286,110 +291,88 @@ def pwrecover(self, *args, **kwargs):
-            skel = self.LostPasswordStep2Skel()
-            if not skel.fromClient(kwargs) or not request.isPostRequest:
-                return self.userModule.render.edit(skel, tpl=self.passwordRecoveryStep2Template)
-            if not securitykey.validate(kwargs.get("skey"), useSessionKey=True):
-                raise errors.PreconditionFailed()
-            self.passwordRecoveryRateLimit.decrementQuota()
-            if not hmac.compare_digest(session["user.auth_userpassword.pwrecover"]["recoveryKey"], skel["recoveryKey"]):
-                # The key was invalid, increase error-count or abort this recovery process altogether
-                session["user.auth_userpassword.pwrecover"]["errorCount"] += 1
-                if session["user.auth_userpassword.pwrecover"]["errorCount"] > 3:
-                    session["user.auth_userpassword.pwrecover"] = None
-                    return self.userModule.render.view(
-                        skel=None,
-                        tpl=self.passwordRecoveryFailedTemplate,
-                        reason=self.passwordRecoveryKeyInvalid)
-
-                return self.userModule.render.edit(skel, tpl=self.passwordRecoveryStep2Template)  # Let's try again
-
-            # If we made it here, the key was correct, so we'd hopefully have a valid user for this
-            user_skel = self.userModule.viewSkel().all().filter(
-                "name.idx =", session["user.auth_userpassword.pwrecover"]["name"]).getSkel()
-
-            if not user_skel:
-                # This *should* never happen - if we don't have a matching account we'll not send the key.
-                session["user.auth_userpassword.pwrecover"] = None
-                return self.userModule.render.view(
-                    skel=None,
-                    tpl=self.passwordRecoveryFailedTemplate,
-                    reason=self.passwordRecoveryUserNotFound)
-
-            if user_skel["status"] != Status.ACTIVE:
-                # The account is locked or not yet validated. Abort the process.
-                session["user.auth_userpassword.pwrecover"] = None
-                return self.userModule.render.view(
-                    skel=None,
-                    tpl=self.passwordRecoveryFailedTemplate,
-                    reason=self.passwordRecoveryAccountLocked)
-
-            # Update the password, save the user, reset his session and show the success-template
-            user_skel["password"] = skel["password"]
-            user_skel.toDB()
-            session["user.auth_userpassword.pwrecover"] = None
-
-            return self.userModule.render.view(None, tpl=self.passwordRecoverySuccessTemplate)
+            recovery_key = utils.generateRandomString(conf["viur.security.password_recovery_key_length"])
+            user_agent = user_agents.parse(current_request.request.headers["User-Agent"])
+            self.sendUserPasswordRecoveryCode(skel["name"], recovery_key, user_agent)  # Send the code in the background
```

Comment:
This line raises 
```py
TypeError: Object of type UserAgent is not JSON serializable
```
`user_agents.parse` creates a `UserAgent` object, this isn't a `dict` as the type hint of `sendUserPasswordRecoveryCode`.